bindings/stronghold-nodejs/src/account/storage/stronghold.rs

```diff
@@ -0,0 +1,167 @@
+    Err(Error::from_reason(
+      "use NapiStronghold.new to instantiate instances".to_owned(),
+    ))
+  }
+
+  /// Creates an instance of `Stronghold`.
+  #[napi]
+  pub async fn new(snapshot: String, password: String, dropsave: Option<bool>) -> Result<NapiStronghold> {
+    Ok(NapiStronghold(
+      Stronghold::new(&snapshot, &*password, dropsave).await.napi_result()?,
+    ))
+  }
+
+  /// Returns whether save-on-drop is enabled.
+  #[napi(getter)]
+  pub fn dropsave(&self) -> bool {
+    self.0.dropsave()
+  }
+
+  /// Set whether to save the storage changes on drop.
+  /// Default: true
+  #[napi(setter)]
+  pub fn set_dropsave(&mut self, dropsave: bool) {
+    self.0.set_dropsave(dropsave);
+  }
+
+  /// Sets the account password.
+  #[napi]
+  pub async fn set_password(&self, password: Vec<u32>) -> Result<()> {
+    let password: [u8; 32] = password
+      .into_iter()
+      .filter_map(|n| u8::try_from(n).ok())
+      .collect::<Vec<u8>>()
+      .try_into()
+      .map_err(|_| Error::from_reason("Invalid password type. Expected [u8; 32]".to_owned()))?;
+    self.0.set_password(password).await.napi_result()
+  }
+
+  /// Write any unsaved changes to disk.
+  #[napi]
+  pub async fn flush_changes(&self) -> Result<()> {
+    self.0.flush_changes().await.napi_result()
+  }
+
+  /// Creates a new keypair at the specified `location`
+  #[napi]
+  pub async fn key_new(&self, did: &NapiDID, location: &NapiKeyLocation) -> Result<String> {
+    let public_key: PublicKey = self.0.key_new(&did.0, &location.0).await.napi_result()?;
+    Ok(encode_b58(&public_key))
+  }
+
+  /// Inserts a private key at the specified `location`.
+  #[napi]
+  pub async fn key_insert(&self, did: &NapiDID, location: &NapiKeyLocation, private_key: String) -> Result<String> {
+    let private_key: PrivateKey = decode_b58(&private_key).napi_result()?.into();
+    let public_key: PublicKey = self
+      .0
+      .key_insert(&did.0, &location.0, private_key)
+      .await
+      .napi_result()?;
+    Ok(encode_b58(&public_key))
+  }
+
+  /// Retrieves the public key at the specified `location`.
+  #[napi]
+  pub async fn key_get(&self, did: &NapiDID, location: &NapiKeyLocation) -> Result<String> {
+    let public_key: PublicKey = self.0.key_get(&did.0, &location.0).await.napi_result()?;
+    Ok(encode_b58(&public_key))
+  }
+
+  /// Deletes the keypair specified by `location`.
+  #[napi]
+  pub async fn key_del(&self, did: &NapiDID, location: &NapiKeyLocation) -> Result<()> {
+    self.0.key_del(&did.0, &location.0).await.napi_result()
+  }
+
+  /// Signs `data` with the private key at the specified `location`.
+  #[napi]
+  pub async fn key_sign(&self, did: &NapiDID, location: &NapiKeyLocation, data: Vec<u32>) -> Result<NapiSignature> {
+    let data_u8: Vec<u8> = data.iter().filter_map(|n| u8::try_from(*n).ok()).collect();
+    if data_u8.len() != data.len() {
+      return Err(Error::from_reason(String::from("Invalid data type. Expected Vec<u8>")));
```

Comment:
Error messages should be completely lowercase if we're following conventions?

```suggestion
      return Err(Error::from_reason(String::from("invalid data type, expected bytes.")));
```

Reply:
That is confusing, yeah. However, we can't take `Vec<u8>` because napi doesn't support that. We also cannot use [Uint8Array](https://docs.rs/napi/latest/napi/bindgen_prelude/struct.Uint8Array.html), unfortunately. See also [this comment](https://github.com/iotaledger/identity.rs/pull/660#discussion_r810070179).

But since the actual reason for a failure here is that we expect bytes, I feel like not saying we expected bytes would be even more confusing?

Reply:
That's so very very unfortunate and odd since `Vec<u32>` is supported.

Yeah I'm fine with saying bytes or some variant if there's no better alternative (which I couldn't think of hence why I couldn't suggest one). I think "bytes" is preferable to `UInt8`/`UInt8Array` because we don't actually expect those data types?

Reply:
There's a whole chain of conversions happening, and if it was TS only, we could probably guarantee that we only get `Uint8Array`s here and the error would become impossible. But JS unfortunately still exists.

Yeah let's go with `Uint8Array` then. Perhaps there should even be a runtime instanceof check in the `keySign` adapter (and others to which this applies) which implements `WasmStorage` and delegates to `Napi`.